### PR TITLE
Try to avoid getting stuck when draining the journal

### PR DESCRIPTION
**Pull request by nalind**
**- What I did**

Try to fix dockerd-keeps-deleted-journal-files-open-forever problems by keeping the reading goroutine from blocking indefinitely.

**- How I did it**

When draining the journal, if the output message channel is full, stop reading, at least for now.  This keeps us from blocking indefinitely when a client hangs up on us and the channel is full.  Doing this requires that we make multiple calls to `drainJournal()`, even in non-follow mode.

In non-follow mode, keep trying to read until we run out of entries or hit the "until" cutoff, which if not specified, we set to "now".  If we don't do that, and we start reading but can't keep up with the journal,
we'll never catch up even after we get to journal messages which were logged long after we started looking for the then-present logs.

The journal API opens the journal files when the journal handle is opened by `sd_journal_open()`, and uses an inotify descriptor to notice when files have been added or removed, but it doesn't set up that
inotify descriptor until the first time that `sd_journal_get_fd()` is called (either by a client, or as part of `sd_journal_wait()`).  We hadn't been doing that until our initial read-through of entries was done,
meaning that we've been missing file deletion events that occurred during our first pass at reading the journal.  Make that window shorter.

Periodically call `sd_journal_process()` when reading the journal, to let it skip over and close handles to journal files which have been deleted since we started reading the journal, for cases where our keeping them open contributes to a shortage of disk space.

Treat `SD_JOURNAL_INVALIDATE` like `SD_JOURNAL_APPEND`, since `sd_journal_process()` prefers to return INVALIDATE over APPEND when both are applicable.  This obsoletes https://github.com/moby/moby/pull/36064, because `sd_journal_process()`'s handling of inotify events should make closing and reopening the handle in response to an INVALIDATE status unnecessary.

Clean up a deferred function call in the journal reading logic.

**- How to verify it**

One way to force a reasonably high turnover rate in the journal is to set `RateLimitBurst=0` and `SystemMaxFileSize=10M` in `/etc/systemd/journald.conf`, and use the `busybox` image running `seq` with a large number (say, a million or ten million) as its argument.

Running `docker logs -f` should follow the container's output until the container exits or the daemon closes the logging handle internally.  `docker logs` should follow along until the container exits or it hits the end of the log.  After the container exits, either command should run to completion and not hang.  If either command is paused at the client (using ctrl-s, waiting several seconds, then ctrl-q to unpause) while running, possibly with an intervening `journalctl --vacuum-size1` pruning down the size of the journal, despite any gaps in the output, the ordering between messages received by the client should be preserved.  If instead of unpausing the client, the client is stopped with ctrl-c, open descriptors that dockerd was using for reading the journal should be closed.

At pretty much any time, we shouldn't be holding open descriptors to deleted journal files for more than a fraction of a second.

**- Description for the changelog**

journald log reader: avoid blocking indefinitely / holding open deleted files

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by thaJeztah** [comment]
ping @cpuguy83 PTAL

File: daemon/logger/journald/read.go [hunk 1]
```diff
@@ -412,22 +516,41 @@ func (s *journald) readLogs(logWatcher *logger.LogWatcher, config logger.ReadCon
...
+		// In case we stopped reading because the output channel was
+		// temporarily full, keep going until we find that we can't
+		// wind forward to the next entry, which we take as an
+		// indication that there are no more entries for us to read, or
+		// until we cross the "until" threshold, or until we get told
+		// to close the reader, whichever comes first.
+		duration := 10 * time.Millisecond
+		timer := time.NewTimer(duration)
+	drainCatchup:
+		for C.sd_journal_next(j) > 0 {
```
Comment (anusha-ragunathan) [inline daemon/logger/journald/read.go]: Shouldn't there be a difference handling EOF (return value 0) and error cases (negative) here?

File: daemon/logger/journald/read.go [hunk 2]
```diff
@@ -412,22 +516,41 @@ func (s *journald) readLogs(logWatcher *logger.LogWatcher, config logger.ReadCon
...
-				// Let followJournal handle freeing the journal context
-				// object and closing the channel.
-				following = true
+			cursor = s.followJournal(logWatcher, j, pipes, cursor, untilUnixMicro)
+			// Let followJournal handle freeing the journal context
+			// object and closing the channel.
+			following = true
+		}
+	} else {
+		// In case we stopped reading because the output channel was
```
Comment (anusha-ragunathan) [inline daemon/logger/journald/read.go]: For my clarification: Did you mean the "output buffer"?

**Comment by anusha-ragunathan** [comment]
:+1: on adding an integration test to cover the leak.

File: daemon/logger/journald/read.go [hunk 3]
```diff
@@ -412,22 +516,41 @@ func (s *journald) readLogs(logWatcher *logger.LogWatcher, config logger.ReadCon
...
+		// temporarily full, keep going until we find that we can't
+		// wind forward to the next entry, which we take as an
+		// indication that there are no more entries for us to read, or
+		// until we cross the "until" threshold, or until we get told
+		// to close the reader, whichever comes first.
+		duration := 10 * time.Millisecond
```
Comment (anusha-ragunathan) [inline daemon/logger/journald/read.go]: Is this an arbitrary frequency to drain?

File: daemon/logger/journald/read.go [hunk 4]
```diff
@@ -178,28 +180,102 @@ func (s *journald) drainJournal(logWatcher *logger.LogWatcher, j *C.sd_journal,
...
+	// Give the journal handle an opportunity to close any open descriptors
+	// for files that have been removed.
+	C.sd_journal_process(j)
+
+	// Seek to the location of the last entry that we sent.
+	if oldCursor != nil {
+		// We know which entry was read last, so try to go to that
+		// location.
+		rc := C.sd_journal_seek_cursor(j, oldCursor)
+		if rc != 0 {
```
Comment (anusha-ragunathan) [inline daemon/logger/journald/read.go]: Nit: Would be nice to stick to one style through this code. Either `rc <0` (Preferable) or `rc != 0`.

File: daemon/logger/journald/read.go [hunk 5]
```diff
@@ -178,28 +180,102 @@ func (s *journald) drainJournal(logWatcher *logger.LogWatcher, j *C.sd_journal,
 	var priority, partial C.int
 	var done bool
 
-	// Walk the journal from here forward until we run out of new entries
-	// or we reach the until value (if provided).
-drain:
+	// Give the journal handle an opportunity to close any open descriptors
+	// for files that have been removed.
+	C.sd_journal_process(j)
```
Comment (anusha-ragunathan) [inline daemon/logger/journald/read.go]: Err check?

File: daemon/logger/journald/read.go [hunk 6]
```diff
@@ -178,28 +180,102 @@ func (s *journald) drainJournal(logWatcher *logger.LogWatcher, j *C.sd_journal,
...
-				if C.sd_journal_next(j) <= 0 {
-					break drain
+		// If we're not keeping up with journald writing to the journal, some of the
+		// files between where we are and "now" may have been deleted since we started
+		// walking the set of entries.  If that's happened, the inotify descriptor in
+		// the journal handle will have pending deletion events after we've been reading
+		// for a while.  Letting the journal library process them will close any that
+		// are already deleted, so that we'll skip over them and allow space that would
+		// have been reclaimed by deleting these files to actually be reclaimed.
+		if sent > 0 && sent%1024 == 0 {
```
Comment (anusha-ragunathan) [inline daemon/logger/journald/read.go]: nit: Since this is garbage collection logic, it'll be easier to read if we can extract this out into a function.

**Comment by thaJeztah** [comment]
ping @nalind could you address the review comments?

**Comment by olljanat** [comment]
Derek add label: rebuild/*

**Comment by olljanat** [comment]
Derek add label: status/failing-ci

**Comment by kolyshkin** [comment]
Just found a more review-able version of this PR here: https://github.com/projectatomic/docker/pull/298. Looking...

**Comment by kolyshkin** [comment]
I proposed to close this one in favor of https://github.com/moby/moby/pull/38859. @nalind what do you think?

**Comment by thaJeztah** [comment]
closing in favour of https://github.com/moby/moby/pull/38859

**Pull request closed without merging**